Proposal for vector db semantic convention

ezimuel opened the pull request:
This is a proposal for vector db semantic convention (see open-telemetry/semantic-conventions-genai#5). I tried to expand the `db` semantic convention adding some `db.vector` attributes. I tried to focus on the basic needs of a general purpose vector database.

I proposed the following experimental attributes (updated with the feedbacks in this PR):
```
db.search.similarity_metric: specify the metric used in similarity search (e.g. cosine)
db.record.id: the ID of the record (e.g. the ID of the vector)
db.vector.field_name: the name field of the vector embedding
db.vector.dimension_count: the dimension of the vector (e.g. 1536)
db.query.limit: the number of vectors returned by a query (e.g. top-k)
```
The operations performed in a vector db, such as `insert`, `update`, `search` and `delete` can be performed using the existing [db.operation.name](https://opentelemetry.io/docs/specs/semconv/attributes-registry/db/) attribute.

Regarding the similarity search we can use the [db.query](https://opentelemetry.io/docs/specs/semconv/attributes-registry/db/) attributes, such as `db.query.parameter.<key>`.

ezimuel committed:
Proposal for vector db semantic convention

gregkalapos commented on model/registry/db.yaml:
```diff
@@ -310,6 +310,10 @@ groups:
               value: 'trino'
               brief: 'Trino'
               stability: experimental
+            - id: vector
```
Isn't just `vector` a bit off in this list?

I'd expect the `db.system` to either one of the well known dbs - like `db.system: elasticsearch` could be used when elasticsearch is used as a vector db. So mostly this will be the specific db system, right?

And then there are other cases, when the db system may not be (yet) on this list. In that case, similar to `other_sql` wouldn't `other_vector` be a better fit?

ezimuel replied:
I agree, `other_vector` is a better fit here.

lmolkova replied:
see my comment on the `db.system` above

lmolkova commented on docs/database/dynamodb.md:
```diff
@@ -83,6 +83,7 @@ These attributes are filled in for all DynamoDB request types.
 | `sybase` | Sybase | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
 | `teradata` | Teradata | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
 | `trino` | Trino | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
+| `vector` | vector | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
```
Many of the databases listed here support vector search. E.g. mongoDB, cosmosDB, cassandra and probably others. A lot of databases are adding vector search due to it's high demand in AI world. Only one `db.system` can be specified on a telemetry item.

So, let's avoid adding a DB feature as a system. Instead, let's define attributes specific to vector search that could be used on any database that supports them.

ezimuel replied:
This means you are in favor to have a generic `vector` attribute to capture a generic vector search not specific to any vendor? This was the goal of this PR, that's why I proposed only few attributes after checking some vendors. But I think the `other_vector` proposed by @gregkalapos makes more sense if the vendor is not listed. The goal is the same.

lmolkova replied:
do we have some real-life case of an instrumentation that does not know the product name of the database?
If so, `other_vector` would make sense.
`other_sql` exists because of JDBC/ODBC/etc drivers and ORMs which might not know what database is used underneath.

If specific databases we need are not listed here, let's just list them and use their names.

ezimuel replied:
There are some proposals for ORM with vector database and some other framework like [Spring AI](https://docs.spring.io/spring-ai/reference/api/vectordbs.html) or [Semantic Kernel](https://learn.microsoft.com/en-us/semantic-kernel/integrations/#available-connectors-to-vector-databases) that are trying to achieve this. But right now I tend to agree, there are no many real-life case.

lmolkova commented on model/registry/db.yaml:
```diff
@@ -529,3 +533,67 @@ groups:
+          The name field as of the vector (e.g. a field name).
+        examples: 'vector'
+      - id: embeddings
+        type: double[]
```
this would be huge and not usable on telemetry - is it necessary?

ezimuel replied:
If you refer to the fact that the embedding values can be huge (eg. 1536) I agree, but I think we need to have a way to represent a vector in the semantic convention. We can mark this as optional.

lmolkova replied:
why do we need to record it on the telemetry? It can be done with logs (and common ways to control verbosity), not as a span attribute

karthikscale3 replied:
+1, it could be recorded with events or logs. Curious to hear what use case drives the requirement to record this?

ezimuel replied:
There's not a specific use case, as semantic convention I would expect to have the embedding values represented in a proposal about vector database (at lest as optional) . Anyway, if you think that it's better to have it in logs for me it's fine.

lmolkova commented on model/registry/db.yaml:
```diff
@@ -529,3 +533,67 @@ groups:
+        brief: >
+          The dimension of the vector.
+        examples: [3]
+      - id: model
```
this should be captured with `gen_ai.request.model` (and other gen-ai attributes) - it's ok to mix different attributes on the same telemetry item.

ezimuel replied:
This is the `model` for the embedding and can be different from the `gen_ai.request.model`. Vector databases are not strictly related with GenAI. There are many use cases where vector db can be used without GenAI (e.g. semantic search).

lmolkova replied:
could you provide some examples of databases that can *compute* embeddings? 

How would database instrumentation know which model was used to create embeddings if it only stores and queries them?

ezimuel replied:
Some vendors that provide in-database embedding:

- [Weaviate](https://weaviate.io/), supporting the following [provider integrations](https://weaviate.io/developers/weaviate/model-providers);
- [Elasticsearch](https://www.elastic.co/), using the [Inference API](https://www.elastic.co/guide/en/elasticsearch/reference/current/inference-apis.html);
- [PostgresML](https://postgresml.org/) provides [in-database embedding generation](https://postgresml.org/docs/open-source/pgml/guides/embeddings/in-database-generation);
- [Vespa](https://vespa.ai/) provides the embedding using the [embed()](https://docs.vespa.ai/en/embedding.html) function;
- [Supabase](https://supabase.com/) provides an embegging generator using [edge function](https://supabase.com/docs/guides/ai/quickstarts/generate-text-embeddings);

I think many other vendors will add this feature. Having the possibility to generate embeddings in the database simplify the customer use cases.

Regarding the question on how database instrumentation know which model was used to create an embedding this is an information that is specified when you create a collection or when you provide some search (e.g. in [PostgreML](https://postgresml.org/docs/open-source/pgml/guides/embeddings/in-database-generation#creating-embeddings)). Moreover, I think the instrumentation libraries can also leverage this `model` attribute since they know the embedding model used to generate the vectors.

lmolkova replied:
It seems all of them use some GenAI model (in-process or external) and provide an integration layer with it. 

I'm not sure what benefit defining a new attribute for databases brings. If embeddings become cross-domain concerns, let's find a generic attribute name for the vectorization model that will be reused between DB and GenAI. For now I strongly recommend to avoid adding new attribute.

lmolkova commented on model/registry/db.yaml:
```diff
@@ -529,3 +533,67 @@ groups:
+        type: int
+        stability: experimental
+        brief: >
+          The dimension of the vector.
```
is it the number of dimensions? let's call it something like `db.vector.dimension_count` and also update the brief

maryliag replied:
@ezimuel you marked this as resolved, but looks like you missed this change, it still shows `dimension` instead of `dimention_count`

ezimuel replied:
Fixed in https://github.com/open-telemetry/semantic-conventions/pull/1231/commits/a3330ffea04ad5d2ce8d9a4bb07a1ce259369339

lmolkova commented on model/registry/db.yaml:
```diff
@@ -529,3 +533,67 @@ groups:
+        type: string
+        stability: experimental
+        brief: >
+          The name field as of the vector (e.g. a field name).
```
do we need both - the name and the id?

ezimuel replied:
The `id` is the identifier of the vector, the `name` is the field name that contains the vector. Many database uses both but some uses only `name` of the field. Maybe we should use a better naming here.

ezimuel replied:
@lmolkova what do you think about this? Thanks.

lmolkova replied:
Could you provide some examples of id and name? 

Let's pick a few (2-3) popular databases and explain what id and name would mean in their context.

E.g. I look into [MongoDB](https://www.mongodb.com/docs/atlas/atlas-vector-search/vector-search-type/#about-the-vector-type) and I don't understand what id or name would be.

Or I look into [Azure Search](https://learn.microsoft.com/en-us/azure/search/vector-store) and don't understand what should go into id.

I look into [pinecone](https://docs.pinecone.io/guides/get-started/quickstart#5-create-a-serverless-index) and it does not talk about ids.

Also, perhaps by vector you mean index?

ezimuel replied:
@lmolkova Here some examples:
- MongoDB uses name for the field (i.e. `path` in search [here](https://www.mongodb.com/docs/atlas/atlas-vector-search/vector-search-stage/#ann-examples) and `field-name` in the definition [here]( https://www.mongodb.com/docs/atlas/atlas-search/field-types/knn-vector/#define-the-index-for-the-fts-field-type-type))
- Azure Search uses name ([here](https://learn.microsoft.com/en-us/azure/search/vector-search-how-to-create-index?tabs=config-2024-07-01%2Crest-2024-07-01%2Cpush%2Cportal-check-index#add-a-vector-field-to-the-fields-collection))
- Pinecone uses id ([here](https://docs.pinecone.io/guides/get-started/quickstart#6-upsert-vectors))
- Qdrant uses id (Point ID, [here](https://qdrant.tech/documentation/concepts/points/#point-ids))
- Elasticsearch uses name ([here](https://www.elastic.co/guide/en/elasticsearch/reference/current/dense-vector.html))
- Milvus uses id ([here](https://milvus.io/docs/insert-update-delete.md#Insert-entities))
- Chroma uses id ([here](https://docs.trychroma.com/getting-started#4.-add-some-text-documents-to-the-collection))
- pgvector for PostgreSQL uses name (i.e. the field name with type vector(x), [here](https://github.com/pgvector/pgvector?tab=readme-ov-file#getting-started))
- Redis uses name (i.e. the field name that will contain the vector values in  `fieldname_embedding`, [here](https://redis.io/docs/latest/develop/get-started/vector-database/#4-generate-text-embeddings))

By **db.vector.name** I mean the **field name** in the document that will be used for embedding (most common for DB like MongoDB, Azure Search, Redis, Elasticsearch, PostgreSQL) and by **db.vector.id** I mean the identifier of the vector (most used in native vector db like Pinecone, Qdrant, Chroma, Milvus). 

Basically these attributes, **name** and **ID**, are used to identify the vector, maybe a better naming for `db.vector.name` can be `db.vector.field_name`?

lmolkova replied:
Thank you for the context!

> name

I agree that `db.vector.field_name` would be more descriptive.
It could be difficult to collect it though - for cosmos, postgres and many other dbs it would require parsing queries (and specific parsing for vector search). I.e. it'd probably mean that very few generic DB instrumentations will do it. 

Even the native instrumentation that we have in CosmosDB does not know if query supplied by user is doing vector search. 

So I wonder how critical this attribute is for the observability purposes. Also I'd be mostly static - is it important enough to justify additional costs of populating it on each span?

A  typical semconv decision making hint is: if we're not sure we need it, let's not add it. Adding attributes to spans is easy, removing them is breaking and hard.

lmolkova replied:
> id

I still don't understand what's behind `db.vector.id` - it seems to be a generic record id and there is nothing vector-specific here. 

I support adding generic attribute like `db.record.id: string` or `db.record.ids: string[]` (needs polishing)

ezimuel replied:
The `db.vector.id` is the record id but specific for the vector. The `db.record.id` will work as well, since it's just an identifier for the record (i.e. vector).

I think we should add `db.vector.field_name` since it is used in many database to specify which field has been used for the embedding.

lmolkova commented on model/registry/db.yaml:
```diff
@@ -529,3 +533,67 @@ groups:
+    brief: >
+      This group defines attributes for vector databases.
+    attributes:
+      - id: similarity
```
suggesting more explicit name like `db.vector.search.similarity_metric`

maryliag replied:
@ezimuel I think you also missed this change, is marked as resolved without any updates :)

ezimuel replied:
Fixed in https://github.com/open-telemetry/semantic-conventions/pull/1231/commits/a3330ffea04ad5d2ce8d9a4bb07a1ce259369339

lmolkova replied:
BTW, who/how/when is going to populate it? 

It seems it's only available at index creation time and not available at query or insert time. I.e. very rarely.

ezimuel replied:
Correct, this is generally provided during the index creation but some databases uses it also in the query (e.g. [Qdrant](https://qdrant.tech/documentation/concepts/search/#metrics)).

I think we can also leverage the [Conditionally Required](https://opentelemetry.io/docs/specs/semconv/general/attribute-requirement-level/) level for some attributes that are not always available, like the `db.vector.search.similarity_metric`, WDYT?

ezimuel replied:
@lmolkova just a reminder for this, thanks.

lmolkova replied:
Thanks, it makes sense! 

BTW I think the attribute should be `db.search.similarity_metric` or similar - similarity is not always based on vectors and we don't need to limit this attribute.

lmolkova commented on model/registry/db.yaml:
```diff
@@ -529,3 +533,67 @@ groups:
+    attributes:
+      - id: similarity
+        type:
+          allow_custom_values: true
```
```suggestion
```

we no longer use this flag, all enums are open

ezimuel replied:
There are still `allow_custom_values: true` in the existing `db.yaml`. Do we need to remove?

karthikscale3 commented:
Thanks for creating this PR. A few additional attributes which we instrument today with our SDK that we have found useful are the following:
`
db.index
db.namespace
db.collection.name
db.top_k
db.query
`

Thoughts on the ones listed above? cc @lmolkova

ezimuel commented:
@karthikscale3 regarding the attributes that you proposed, some already exists:
- `db.index` can be represented using `db.operation.name = index`;
- `db.namespace` already exists;
- `db.collection.name` already exists;
- `db.query` can be represented using `db.query.text` and also `db.query.parameter`;

The `top_k` proposal I think it's a good idea:
- `db.vector.query.top_k` to represent the k most similar vectors returned by a similarity serach;

Moreover, I found very interesting the proposal of [OpenLLMetry](https://github.com/traceloop/openllmetry) project especially the part regarding the attributes for vector db, [here](https://github.com/traceloop/openllmetry):
```
 # Vector DB
 VECTOR_DB_VENDOR = "db.system"
 VECTOR_DB_OPERATION = "db.operation"
 VECTOR_DB_QUERY_TOP_K = "db.vector.query.top_k"
```

ezimuel committed:
Merge + applied feedbacks #1231

ezimuel commented:
@lmolkova I applied all the feedbacks, thanks for the review. @karthikscale3 I added the `top_k` attribute, thanks.

Summary of the changes:
- removed `vector` in `db.system`;
- removed the `db.vector.embeddings`;
- renamed `db.vector.dimension` in `db.vector.dimension_count`;
- added the `db.vector.query.top_k` as suggested by @karthikscale3;

ezimuel committed:
Removed allow_custom_values: true in db.yaml

karthikscale3 commented:
> @karthikscale3 regarding the attributes that you proposed, some already exists:
> 
> * `db.index` can be represented using `db.operation.name = index`;
> * `db.namespace` already exists;
> * `db.collection.name` already exists;
> * `db.query` can be represented using `db.query.text` and also `db.query.parameter`;
> 
> The `top_k` proposal I think it's a good idea:
> 
> * `db.vector.query.top_k` to represent the k most similar vectors returned by a similarity serach;
> 
> Moreover, I found very interesting the proposal of [OpenLLMetry](https://github.com/traceloop/openllmetry) project especially the part regarding the attributes for vector db, [here](https://github.com/traceloop/openllmetry):
> 
> ```
>  # Vector DB
>  VECTOR_DB_VENDOR = "db.system"
>  VECTOR_DB_OPERATION = "db.operation"
>  VECTOR_DB_QUERY_TOP_K = "db.vector.query.top_k"
> ```

Yea that sounds good! And yes, my intention was to reuse the existing ones. Wasn't sure if we needed them redefined for the sake of vector dbs or not. But sounds like its unnecessary.

karthikscale3 commented:
> @lmolkova I applied all the feedbacks, thanks for the review. @karthikscale3 I added the `top_k` attribute, thanks.
> 
> Summary of the changes:
> 
> * removed `vector` in `db.system`;
> * removed the `db.vector.embeddings`;
> * renamed `db.vector.dimension` in `db.vector.dimension_count`;
> * added the `db.vector.query.top_k` as suggested by @karthikscale3;
> * removed the `allow_custom_values: true` in `db.yaml`, see last commit;

Thank you! From my side, everything looks good. We discussed this PR in today's working group call and @nirga wanted to take a deeper look at it once again.

trask commented on docs/attributes-registry/db.md:
```diff
@@ -6,12 +6,22 @@
 
 # Db
 
+<<<<<<< HEAD
```
looks like a bad merge

trask commented on docs/database/dynamodb.md:
```diff
@@ -9,7 +9,90 @@ linkTitle: AWS DynamoDB
 The Semantic Conventions for [AWS DynamoDB](https://aws.amazon.com/dynamodb/) extend and override the general
 [AWS SDK Semantic Conventions](/docs/cloud-providers/aws-sdk.md) and [Database Semantic Conventions](database-spans.md).
 
+<<<<<<< HEAD
```
and here

ezimuel committed:
Fixed merge

ezimuel commented:
I fixed the merge issues. Thanks @trask

ezimuel committed:
Merge branch 'main' into vector-db

nirga reviewed (commented):
Thanks, that's a great start! I wonder if we want to add specific spans that use these attributes in this PR as well?

ezimuel commented:
@nirga can you give me an example of specific span? FYI, I'm going offline and I'll come back August 4 for further discussion.

nirga commented:
Sorry, nvm I think this is already covered as part of the DB semconv

maryliag commented on docs/attributes-registry/db.md:
```diff
@@ -199,6 +200,28 @@ This group defines attributes for Elasticsearch.
 
 **[8]:** Many Elasticsearch url paths allow dynamic values. These SHOULD be recorded in span attributes in the format `db.elasticsearch.path_parts.<key>`, where `<key>` is the url path part name. The implementation SHOULD reference the [elasticsearch schema](https://raw.githubusercontent.com/elastic/elasticsearch-specification/main/output/schema/schema.json) in order to map the path part values t ...
 
+## Db Vector Attributes
```
nit: Vector Database Attributes

maryliag commented:
@ezimuel looks like you missed some of the changes you marked as resolved:
- rename `db.vector.dimension` to `db.vector.dimension_count`: it is still showing `db.vector.dimension`
- rename `db.vector.similarity` to `db.vector.search.similarity_metric`: it is still showing `db.vector.similarity`

lmolkova commented on model/registry/db.yaml:
```diff
@@ -533,3 +530,66 @@ groups:
+        brief: >
+          The model used for the embedding.
+        examples: 'text-embedding-3-small'
+      - id: query.top_k
```
I think we should come up with a more common attribute not specific to vector dbs. 
Many databases allow to limit number of returned rows:
- JDBC has [Statement.setMaxRows](https://docs.oracle.com/javase/7/docs/api/java/sql/Statement.html#setMaxRows(int)), 
- Mongo allows to set a [limit](https://www.mongodb.com/docs/drivers/java/sync/current/fundamentals/crud/read-operations/limit/#specify-a-limit)

Suggesting `db.query.max_returned_items` or similar

ezimuel replied:
@lmolkova I see the similarity here but I think the `db.vector.query.top-k` is more specific from a semantic point of view and more related to vectors, since it specifies the top k results in order, starting from the most similar. In semantic search we have this **similarity** value that is always present in any result that we don't have in standard database. The `limit` parameter of SQL returns the first k results but not in order, it depends on how you build the query (e.g. using `ORDER BY`).
 I personally think we should keep `db.vector.query.top-k` and potentially add a `db.query.limit` (or `db.query.max_returned_items` as you suggested) in a separate PR.

lmolkova replied:
Out of the few dbs I checked, they use `limit` in vector search
- pgvector uses [traditional `limit`](https://github.com/pgvector/pgvector?tab=readme-ov-file#getting-started) - same with cosmos and other sql databases
- mongo uses [imit](https://www.mongodb.com/docs/atlas/atlas-vector-search/tutorials/vector-search-quick-start/#std-label-vector-search-quick-start)
- qdrant uses [limit](https://qdrant.tech/documentation/concepts/search/#search-api)


So we're saying that DB instrumentations will need to detect if query is related to vector search or not and depending on this populate top-k or limit. That's difficult or impossible, but most importantly inconsistent and depends on instrumentation capabilities.

I.e. instrumentations that don't have vector-db specifics and those that do will use different attributes for the same thing.

So, I'd still prefer `db.query.limit` or something similar (and it should be under the same condition as `db.query.text` - we cannot require instrumentations to do query parsing)

ezimuel replied:
@lmolkova do you agree that `top-k` and `limit` are two different concepts, based on my previous comment? If they are I think we cannot use a single attribute (e.g. `db.query.limit`) to manage both.

ezimuel replied:
@lmolkova just a reminder for this, thanks.

lmolkova replied:
@ezimuel I see that databases use both terms to describe the same thing (see my comment above).

Let's say you have a postgres query like `SELECT * FROM items ORDER BY embedding <-> '[3,1,2]' LIMIT 5;` - the general-purpose DB instrumentation can report `limit`. If you make it understand vector search syntax, it may be able to use `top_k` instead, but that's would be inconsistent and unfamiliar for those who use vector search in postgres.

ezimuel replied:
@lmolkova I see your point but I think `top-k` has a different meaning from `limit`. If you are using a relation database as vector db, `limit` is fine since you are building an SQL statement and you specify the order. But, if you are using a native vector database (e.g. Qdrant), the `top-k` is more relevant since `top` implies the order, using a similarity metric.

I think we should add both:
- `db.query.limit`
- `db.vector.query.top-k`

ezimuel replied:
@lmolkova I found that the `SELECT LIMIT` is not part of the SQL standard. A statement compliant with SQL standard is `FETCH FIRST`. Moreover, I discovered that we already have `db.operation.parameter,<key>` in the semantic convention. This means, we can have `db.operation.parameter.fetch-first`. I don't think we need to add a `limit` or `fetch-first` in `db.query`. What do you think?

At the same time, I believe we should include `db.vector.query.top-k`. As noted earlier, it differs in meaning from `fetch-first` and, additionally, this parameter is much easier to retrieve in a vector database because it's not part of any statement.

lmolkova commented:
We need to reference new attributes in the database spans conventions (see https://github.com/open-telemetry/semantic-conventions/blob/main/model/trace/database.yaml), specifically on the conventions for the databases we have there which support vector search. 

We should describe how new attributes apply to them.

ezimuel committed:
Merge remote-tracking branch 'upstream/main' into vector-db

ezimuel committed:
Updated dimension_count and similarity_metric

ezimuel committed:
Merge remote-tracking branch 'origin/vector-db' into vector-db

ezimuel commented:
> @ezimuel looks like you missed some of the changes you marked as resolved:
> 
> * rename `db.vector.dimension` to `db.vector.dimension_count`: it is still showing `db.vector.dimension`
> * rename `db.vector.similarity` to `db.vector.search.similarity_metric`: it is still showing `db.vector.similarity`

I just fixed it, thanks @maryliag

ezimuel committed:
Merge branch 'main' into vector-db

Commit:
Fix array attribute examples  (#1325)

lmolkova commented:
> Thanks for creating this PR. A few additional attributes which we instrument today with our SDK that we have found useful are the following:

> db.index
> db.namespace
> db.collection.name
> db.top_k
> db.query

The index name I believe should be captured in the `db.collection.name` - we store and query against this collection/index

As @ezimuel already pointed out some are already defined (except top-k) and we have discussions around it (https://github.com/open-telemetry/semantic-conventions/pull/1231#discussion_r1710484074).

ChrsMark committed:
Add k8s.{pod,node}.cpu.{time,usage} metrics (#1320)

Signed-off-by: ChrsMark <chrismarkou92@gmail.com>

maryliag committed:
Db metrics pending requests (#1290)

Co-authored-by: Liudmila Molkova <limolkova@microsoft.com>

Commit:
Fix `process.args_count` attribute (#1331)

ChrsMark committed:
Add k8s.volume.{name,type} attributes (#1251)

Signed-off-by: ChrsMark <chrismarkou92@gmail.com>

MadVikingGod committed:
Add tests for rego policies (#1334)

Co-authored-by: Aaron Clawson <MadVikingGod@users.noreply.github.com>
Co-authored-by: Liudmila Molkova <limolkova@microsoft.com>

maryliag committed:
add `nodejs.eventloop.time` metric (#1259)

Co-authored-by: Liudmila Molkova <limolkova@microsoft.com>

MSNev committed:
chore: Remove support for the event `fields` referencing/inheriting definitions from global attributes. (#1340)

jsuereth committed:
Attempt to optimise attribute name collision checks. (#1328)

Co-authored-by: Liudmila Molkova <limolkova@microsoft.com>

jsuereth committed:
(chore) Add dependabot config to keep tooling up to date. (#1346)

Co-authored-by: Armin Ruech <7052238+arminru@users.noreply.github.com>

ChrsMark committed:
Fix broken docker link (#1332)

Signed-off-by: ChrsMark <chrismarkou92@gmail.com>
Co-authored-by: Liudmila Molkova <limolkova@microsoft.com>

dependabot[bot] committed:
Bump markdownlint-cli from 0.31.0 to 0.41.0 (#1349)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Liudmila Molkova <limolkova@microsoft.com>

dependabot[bot] committed:
Bump go.opentelemetry.io/build-tools/chloggen from 0.12.0 to 0.14.0 in /internal/tools (#1350)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Liudmila Molkova <limolkova@microsoft.com>

dependabot[bot] committed:
Bump gulp from 4.0.2 to 5.0.0 (#1348)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Commit:
Fix link anchors (#1354)

maryliag committed:
chore: update ids (#1352)

Co-authored-by: Liudmila Molkova <limolkova@microsoft.com>

ezimuel committed:
Removed db.vector.id and added db.record.id, renamed db.vector.field_name

ezimuel commented:
@lmolkova I provided the following changes:
- renamed `db.vector.name` in `db.vector.field_name`;
- removed `db.vector.id` in favor of `db.record.id`.

I think the only missing point is about `db.vector.query.top-k`, see my [last comment](https://github.com/open-telemetry/semantic-conventions/pull/1231#discussion_r1722969696).
Thanks!

ezimuel committed:
Merge branch 'main' into vector-db

lmolkova commented:
thanks for working on this @ezimuel !

Please make sure to update actual database semantic conventions and reference attributes you're adding under those that report them - https://github.com/open-telemetry/semantic-conventions/pull/1231#issuecomment-2251896744

ezimuel committed:
Merge from upstream/main

ezimuel committed:
Removed db.vector.model and moved db.vector.search.similarity_metric in db.search.similarity_metric

ezimuel commented:
@AlexanderWert I updated the PR and applied the suggestions from @lmolkova. I did the following changes:
- removed the `db.vector.model` as reported [here](https://github.com/open-telemetry/semantic-conventions/pull/1231#discussion_r1771747809);
- moved `db.vector.search.similarity_metric` to `db.search.similarity_metric`, as suggested [here](https://github.com/open-telemetry/semantic-conventions/pull/1231#discussion_r1771733525);

The only open question is about `limit` vs `top-k`, see my [last comment](https://github.com/open-telemetry/semantic-conventions/pull/1231#discussion_r1775087483).

@lmolkova I didn't understand what I supposed to do in [this comment](https://github.com/open-telemetry/semantic-conventions/pull/1231#issuecomment-2251896744), since the link https://github.com/open-telemetry/semantic-conventions/blob/main/model/trace/database.yaml does not work. Can you clarify? Thanks.

ezimuel committed:
Merge branch 'main' into vector-db

lmolkova commented:
@ezimuel  I'm sorry I did not reply earlier. I'm swamped with some work at the moment and, unfortunately, it might take me some time to reply.

If you look into https://github.com/open-telemetry/semantic-conventions/tree/main/docs/database you'd see that we have some docs for individual database systems - we reference attributes from the registry and we explain how these attributes are applied to this system (or not applied). 

This is powered by the yaml in https://github.com/open-telemetry/semantic-conventions/tree/main/model/database.

Please look at the existing database conventions and update them to include new attributes you're adding.

ezimuel commented:
@lmolkova thanks for the information and sorry also on my side for the late reply, very busy period.
I'm looking into the examples that you shared and I'll send a commit soon.

ezimuel committed:
Merge remote-tracking branch 'upstream/main' into vector-db

ezimuel committed:
Merge remote-tracking branch 'upstream/main' into vector-db

ezimuel committed:
Merge branch 'vector-db' of github.com:ezimuel/semantic-conventions into vector-db

ezimuel committed:
Merge remote-tracking branch 'upstream/main' into vector-db

ezimuel committed:
Merge remote-tracking branch 'upstream/main' into vector-db

ezimuel committed:
Added docs + spans

ezimuel commented:
@lmolkova and @AlexanderWert I finally found some time to work on this PR. I’ve added the spans definition and updated the documentation. I hope everything looks good. Looking forward to your review. Thanks!

ezimuel committed:
Merge remote-tracking branch 'upstream/main' into vector-db

AlexanderWert commented on model/database/spans.yaml:
```diff
@@ -764,3 +764,81 @@ groups:
         requirement_level:
           conditionally_required: If available.
+
+  - id: span.db.vector.client
```
agree, but how about doing it in a follow up issue / PR, as it is additive information?

lmolkova replied:
I don't think we should introduce `span.db.vector.client` span - it's not clear what/when/how will report it. 
Also, referencing attributes in the specific conventions is a good test if there attributes are applicable and properly named/described to fit to specific db implementations.

E.g. `db.vector.query.top_k` debate will become obvious on postgress - you can't have an attribute like this there and would have to call it `limit`. 

It's ok to limit the scope to several DBs and not cover all possible details.

AlexanderWert replied:
> I don't think we should introduce `span.db.vector.client` span

Sorry, I don't get that. We also have a generic [`span.db.client`](https://github.com/open-telemetry/semantic-conventions/blob/41053b507998b6dc4effa22f657cea96ab5f8fe8/model/database/spans.yaml#L111) (which then in addition has technology specific overrides). Wouldn't it be the same thing here, so having a generic description for vector db client spans + in addition having technology-specific overrides for the DBs listed above.

> it's not clear what/when/how will report it

agree, in the readme above we need more explicit / clearer guidance on when to use `span.db.vector.client` vs. the more general `span.db.client`, but we should do it in a technology-independent way first.

> Also, referencing attributes in the specific conventions is a good test if these attributes are applicable and properly named/described to fit to specific db implementations.
> ...
> It's ok to limit the scope to several DBs and not cover all possible details.

Makes sense.

ezimuel replied:
I agree with @AlexanderWert . I think we should have a generic `span.db.vector.client` for vector databases that do not have technology-specific conventions, like the generic sql clients. 
@lmolkova  I see the issue with `db.vector.query.top_k` and `db.query.limit`, I think I can remove the `top-k` even if they are semantically different.

lmolkova replied:
> Sorry, I don't get that. We also have a generic [span.db.client](https://github.com/open-telemetry/semantic-conventions/blob/41053b507998b6dc4effa22f657cea96ab5f8fe8/model/database/spans.yaml#L111) (which then in addition has technology specific overrides). Wouldn't it be the same thing here, so having a generic description for vector db client spans + in addition having technology-specific overrides for the DBs listed above.

The problem is that none of the databases we have in the semconv should be considered a vector databases then. The  `span.db.vector.client` is an abstract thing nothing is going to use. MongoDB, CosmosDB, elastic, postres - all of them are used as vector databases - will they extend `span.db.vector.client`? Will generic instrumentations for them cover vector-related operations?

Also we're trying to introduce attributes without attempting to apply them to existing databases - this is the major blocker I see in this PR.

AlexanderWert commented on model/database/spans.yaml:
```diff
@@ -764,3 +764,81 @@ groups:
+        requirement_level: recommended
+        examples: [3, 1536]
+      - ref: db.vector.query.top_k
+        brief: the top-k most similar vectors returned by a query
```
```suggestion
        brief: The top-k most similar vectors returned by a query
```

AlexanderWert commented on model/database/spans.yaml:
```diff
@@ -764,3 +764,81 @@ groups:
+          conditionally_required: If available.
+        examples: ['image_vector', 'embedding_field']
+      - ref: db.vector.dimension_count
+        brief: the dimension of the vector (e.g. 1536)
```
```suggestion
        brief: The dimension of the vector (e.g. 1536)
```

AlexanderWert commented on model/database/spans.yaml:
```diff
@@ -764,3 +764,81 @@ groups:
+          conditionally_required: If available.
+        examples: ['1', '5c56c793-69f3-4fbf-87e6-c4bf54c28c26']
+      - ref: db.vector.field_name
+        brief: the name field of the vector embedding
```
```suggestion
        brief: The name field of the vector embedding
```

AlexanderWert commented on model/database/spans.yaml:
```diff
@@ -764,3 +764,81 @@ groups:
+          conditionally_required: If available.
+        examples: ['cosine', 'dot', 'euclidean', 'manhattan']
+      - ref: db.record.id
+        brief: the ID of the record (e.g. the ID of the vector)
```
```suggestion
        brief: The ID of the record (e.g. the ID of the vector)
```

AlexanderWert commented on model/database/spans.yaml:
```diff
@@ -764,3 +764,81 @@ groups:
+          If the query doesn't target a specific index, this field MUST NOT be set.
+        examples: ['my_index', 'index1, index2']
+      - ref: db.search.similarity_metric
+        brief: specify the metric used in similarity search (e.g. cosine)
```
```suggestion
        brief: Specifies the metric used in similarity search (e.g. cosine)
```

ezimuel committed:
Update model/database/spans.yaml

Co-authored-by: Alexander Wert <AlexanderWert@users.noreply.github.com>

ezimuel committed:
Update model/database/spans.yaml

Co-authored-by: Alexander Wert <AlexanderWert@users.noreply.github.com>

ezimuel committed:
Update model/database/spans.yaml

Co-authored-by: Alexander Wert <AlexanderWert@users.noreply.github.com>

ezimuel committed:
Update model/database/spans.yaml

Co-authored-by: Alexander Wert <AlexanderWert@users.noreply.github.com>

ezimuel committed:
Update model/database/spans.yaml

Co-authored-by: Alexander Wert <AlexanderWert@users.noreply.github.com>

ezimuel commented:
Thanks @AlexanderWert for the feedback.

ezimuel committed:
Merge remote-tracking branch 'upstream/main' into vector-db

ezimuel committed:
Merge branch 'vector-db' of github.com:ezimuel/semantic-conventions into vector-db

TaoChenOSU commented on docs/attributes-registry/db.md:
```diff
@@ -228,9 +233,38 @@ This group defines attributes for Elasticsearch.
+
+| Value  | Description | Stability |
+|---|---|---|
+| `cosine` | The cosine metric. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
```
nit: 
1. There are cosine similarity and cosine distance. Better to distinguish the two in the convention.
2. There are a few common ones that are not listed here: Squared Euclidean and hamming.

TaoChenOSU commented on docs/attributes-registry/db.md:
```diff
@@ -228,9 +233,38 @@ This group defines attributes for Elasticsearch.
+---
 
-**[11] `db.elasticsearch.path_parts`:** Many Elasticsearch url paths allow dynamic values. These SHOULD be recorded in span attributes in the format `db.elasticsearch.path_parts.<key>`, where `<key>` is the url path part name. The implementation SHOULD reference the [elasticsearch schema](https://raw.githubusercontent.com/elastic/elasticsearch-specification/main/output/schema/schema.json) in orde ...
+`db.search.similarity_metric` has the following list of well-known values. If one of them applies, then the respective value MUST be used; otherwise, a custom value MAY be used.
```
The similarity metric doesn't need to be defined per search request. It can also be defined when the collection is created.

TaoChenOSU commented on docs/attributes-registry/db.md:
```diff
@@ -228,9 +233,38 @@ This group defines attributes for Elasticsearch.
+
+| Attribute | Type | Description | Examples | Stability |
+|---|---|---|---|---|
+| <a id="db-vector-dimension-count" href="#db-vector-dimension-count">`db.vector.dimension_count`</a> | int | The dimension of the vector. | `3` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
```
nit: `db-vector-dimension-count` -> `db-vector-dimension`.

lmolkova commented on docs/database/vector.md:
```diff
@@ -0,0 +1,197 @@
+| Attribute  | Type | Description  | Examples  | [Requirement Level](https://opentelemetry.io/docs/specs/semconv/general/attribute-requirement-level/) | Stability |
+|---|---|---|---|---|---|
+| [`db.operation.name`](/docs/attributes-registry/db.md) | string | The operation  to be performed on the vector database (e.g. build an index/collection) [1] | `build`; `insert`; `search`; `delete` | `Required` | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
+| [`http.request.method`](/docs/attributes-registry/http.md) | string | HTTP request method. [2] | `GET`; `POST`; `HEAD` | `Required` | ![Stable](https://img.shields.io/badge/-stable-lightgreen) |
```
`http.request.method` should not be used on generic database span
